src/hotspot/share/opto/macro.cpp

```diff
@@ -2569,3 +2591,35 @@ bool PhaseMacroExpand::expand_macro_nodes() {
   _igvn.set_delay_transform(false);
   return false;
 }
+
+#ifndef PRODUCT
+int PhaseMacroExpand::_objs_scalar_replaced_counter = 0;
+int PhaseMacroExpand::_monitor_objects_removed_counter = 0;
+int PhaseMacroExpand::_GC_barriers_removed_counter = 0;
+int PhaseMacroExpand::_memory_barriers_removed_counter = 0;
+
+void PhaseMacroExpand::print_statistics() {
+  tty->print("Objects scalar replaced = %d, ", Atomic::load(&_objs_scalar_replaced_counter));
+  tty->print("Monitor objects removed = %d, ", Atomic::load(&_monitor_objects_removed_counter));
+  tty->print("GC barriers removed = %d, ", Atomic::load(&_GC_barriers_removed_counter));
+  tty->print_cr("Memory barriers removed = %d", Atomic::load(&_memory_barriers_removed_counter));
+}
+
+int PhaseMacroExpand::count_MemBar() {
```

Comment:
I am not sure about this procedural. Even though you use Unique_Node_List, is it still possible to count the same membar multiple times?  or a backedge causes a infinitely loop? 

I think you can use Compile::identify_useful_nodes to collect all useful nodes and then count MemBar() nodes.

Reply:
@navyxliu Since no nodes are ever removed from the list, I think this issue is avoided. If we find a loop back to a node we have already searched and attempt to push this node at line 2620, it will not be added to the list because it already exists.

Reply:
nits: 
This function seems not to be part of 'PhaseMacroExpand', at least it could be a static member function.

Reply:
You should check `PrintOptoStatistic` flag.